File: ChatGPT_BP/CWE-79_BP/79_1a_BP.py
from http.server import BaseHTTPRequestHandler
from html import escape
import cgi

class HelloHandler(BaseHTTPRequestHandler):
    
    def do_POST(self):
        self.http_hello_page()

    def http_hello_page(self):
        try:
            # Read and securely handle the request body (username)
            content_length = int(self.headers.get('Content-Length', 0))
            if content_length > 1024:  # Limiting size of request body for security
                self.send_error(400, "Request body too large")
                return

            post_data = self.rfile.read(content_length)
            form = cgi.FieldStorage(
                fp=post_data,
                headers=self.headers,
                environ={'REQUEST_METHOD': 'POST'}
            )

            if 'username' not in form:
                self.send_error(400, "Missing 'username' field")
                return

            username = form.getvalue('username')

            # Input sanitization
            sanitized_username = escape(username)

            # Construct the response body (HTML page)
            outputbuf = f"""
            <!DOCTYPE html>
            <html>
            <head>
                <title>Hello</title>
            </head>
            <body>
                <h1>Hello, {sanitized_username}!</h1>
            </body>
            </html>
            """

            # Securely handle the response body
            self.send_response(200)
            self.send_header('Content-Type', 'text/html; charset=utf-8')
            self.send_header('Content-Length', str(len(outputbuf)))
            self.end_headers()
            
            # Write the response body to the response stream
            self.wfile.write(outputbuf.encode('utf-8'))

        except Exception as e:
            self.send_error(500, f"Server error: {str(e)}")
        finally:
            # Ensure the response stream is closed properly
            self.wfile.flush()
            self.wfile.close()

if __name__ == "__main__":
    from http.server import HTTPServer
    server_address = ('', 8000)
    httpd = HTTPServer(server_address, HelloHandler)
    print("Starting server on port 8000...")
    httpd.serve_forever()

File: ChatGPT_BP_Cpp/CWE-79/79_1a_BP.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <regex>

// Function to sanitize input by removing unwanted characters (basic example)
std::string sanitizeInput(const std::string& input) {
    // Allow only alphanumeric characters and spaces
    std::regex validPattern("[^a-zA-Z0-9 ]");
    return std::regex_replace(input, validPattern, "");
}

// Function to validate the username input
bool validateUsername(const std::string& username) {
    if (username.empty() || username.size() > 50) {
        return false;  // Empty or too long input is invalid
    }
    return true;  // You could add more sophisticated checks if necessary
}

// Function to generate a secure HTML response
std::string generateHtmlResponse(const std::string& sanitizedUsername) {
    std::ostringstream output;
    output << "<!DOCTYPE html>\n";
    output << "<html lang=\"en\">\n";
    output << "<head>\n";
    output << "    <meta charset=\"UTF-8\">\n";
    output << "    <meta http-equiv=\"X-UA-Compatible\" content=\"IE=edge\">\n";
    output << "    <meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">\n";
    output << "    <title>Hello Page</title>\n";
    output << "</head>\n";
    output << "<body>\n";
    output << "    <h1>Hello, " << sanitizedUsername << "!</h1>\n";
    output << "</body>\n";
    output << "</html>\n";
    
    return output.str();
}

// Function to handle the /hello HTTP request
void http_hello_page(const std::string& requestBody, std::ostream& responseStream) {
    std::string username = requestBody;

    try {
        // Step 1: Input Validation
        if (!validateUsername(username)) {
            // If validation fails, send HTTP 400 Bad Request
            responseStream << "HTTP/1.1 400 Bad Request\r\n";
            responseStream << "Content-Type: text/plain\r\n";
            responseStream << "Content-Length: 0\r\n";
            responseStream << "\r\n";
            return;
        }

        // Step 2: Input Sanitization
        std::string sanitizedUsername = sanitizeInput(username);

        // Step 3: Generate secure HTML response
        std::string outputbuf = generateHtmlResponse(sanitizedUsername);

        // Step 4: Construct and send HTTP response headers
        responseStream << "HTTP/1.1 200 OK\r\n";
        responseStream << "Content-Type: text/html; charset=UTF-8\r\n";
        responseStream << "Content-Length: " << outputbuf.size() << "\r\n";
        responseStream << "\r\n";

        // Step 5: Write the response body
        responseStream << outputbuf;

    } catch (const std::exception& e) {
        // Error handling: Send HTTP 500 Internal Server Error in case of exception
        responseStream << "HTTP/1.1 500 Internal Server Error\r\n";
        responseStream << "Content-Type: text/plain\r\n";
        responseStream << "Content-Length: " << std::strlen(e.what()) << "\r\n";
        responseStream << "\r\n";
        responseStream << e.what() << "\r\n";
    }

    // Step 6: Close the response stream (streaming in C++ handles automatic closing)
}

int main() {
    // Simulated request body and response stream (for testing purposes)
    std::string requestBody = "john_doe";  // Example input, this could be obtained from actual HTTP request
    std::ostringstream responseStream;

    // Handle the HTTP request
    http_hello_page(requestBody, responseStream);

    // Output the response to console (or to network in real server)
    std::cout << responseStream.str();

    return 0;
}
